### ISSUE TITLE
Question about reduce_by_key

### ISSUE THREAD

**Issue by gcarvajalb** [ddemidov/vexcl#199]
Hello,

I'm moving to VexCL from Bolt, after finding out that VexCL was much more versatil. So far so good, but I'm finding a little issue with an specific use of the reduce_by_key function.

I'm calling reduce_by_key in a loop, and each iteration removes an element of the input values and keys. With Bolt I had something like this:

```cpp
int keys[N] ;
int vals[N];
int keys_out[N];
int vals_out[N];
for (i =0; i<N; i++){
           DosomecomputewithKeysandVals();
           bolt::cl::reduce_by_key( keys+i, keys+N, vals+i, keys_out, vals_out );
}
```

So the length of the vectors to process is reduced with each iteration. 

From what I've tried, the VexCL version of reduce_by_key doesn't allow passing and offset to the input vectors, so I have something like this now:

```cpp
int keys[N] ;
int vals[N];
int keys_out[N];
int vals_out[N];
for(i= 0; i<n; i++ ){
      DosomecomputewithKeysandVals();
      Put_a_negative_key_in_keys[i]();
      num_keys_max = vex::reduce_by_key(
        	keys, vals, //d_ikey, d_ivals,
            	keys_out, vals_out,
            	equal, maximum
        	);
    discard_vals_out_with_negative_keys(); 
}
```

Thus in every iteration I pass the entire vector, masking the elements I don't need with a negative key to discard them later. It works acceptably good (still see a significant acceleration with our reference C version), but obviously is not really efficient. On the other hand, having the precise number of valid keys in the output vectors solved some headaches I was having with Bolt.

Is there any way of passing the input vectors with an offset, similar to what Bolt offers?

I'm also considering the slicing functions of VexCL, but since the vector length changes in every iteration guess it will be overkill.

I would appreciate any hint or recommendation.

**Comment by ddemidov** [comment]
I am afraid there is no way to do this efficiently with VexCL. The `reduce_by_key()` algorithm was [adapted from Bolt](https://github.com/ddemidov/vexcl/blob/master/vexcl/reduce_by_key.hpp#L33), so it has the same [poor performance](http://arxiv.org/abs/1601.03144). It also has reduced functionality, as it always takes vectors of tuples of vectors as input/output, so using slices won't work in this case.

A possible solution would be to use the [Boost.Compute](http://vexcl.readthedocs.org/en/latest/interop.html) backend with VexCL, and then use the Boost.Compute's version of [reduce_by_key()](http://www.boost.org/doc/libs/develop/libs/compute/doc/html/boost/compute/reduce_by_key.html).

**Comment by gcarvajalb** [comment]
Hi,
that confirms my suspicion, so thanks for the quick reply and saving me some time of scratching my head. 

The inefficient version I have right now is still good enough to solve an immediate bottleneck we were having in our application, but I'd like to get the most juice of it. I'm actually trying with the Boost.compute version of reduce_by_key(), but need to do some extra work to get the number of keys, which VexCL provides as the output. I'll spend sometime evaluating the trade-offs.

Also, given it's my first communication on the topic, I take advantage of this message to acknowledge the terrific job with the library. I was just about to ditch my OpenCL adventure after some frustrating entry experiences (had some previous experience with CUDA) until I found your articles and paper.

Cheers

**Comment by ddemidov** [comment]
Thanks!

It looks like Boost.Compute's version returns a pair of iterators to the end of the key and value ranges, so you could easily get the number of outputs e.g. by `std::distance(keys_result, ret.first)` (where `ret` is the return value of the `readuce_by_key()`).

**Comment by gcarvajalb** [comment]
Neat!

Thanks for the tip. I've missed the return value from boost.compute's reduce_by_key().
It was kind of hidden, but in case someone else need it, the precise details are here (which is exactly what Denis mentioned):  
https://github.com/boostorg/compute/blob/master/perf/perf_reduce_by_key.cpp

Functionally is working good enough for our immediate purpose. I'll dig deeper into the performance evaluation later though.

After some initial frustration, OpenCL is becoming kind of fun now :).

Thanks again for the quick reply and the spot-on tips.

**Comment by gcarvajalb** [comment]
Well, after some experimentation I managed to have an equivalent implementation of my VexCL-based program using boost::compute.

At first sight and after some quick informal tests, the VexCL version of reduce_by_key seems to be way faster than the boost::compute's version. But what is really weird is that, considering the following loop:

```
for (i =0; i<N; i++){
           boost::compute::reduce_by_key( keys.begin()+i, keys.end(), vals.begin(), 
                                                                keys_out, vals_out );
}
```
the time taken on each iteration increases with the value of i, meaning the algorithm takes longer with less elements. For some reason, looks like there is some sort of penalization when adding an offset to the iterators.

I've more experience with low-level C and I'm kind of new to the C++ concepts (iterators, vector type, etc.) and also to OpenCL, so I may be missing something fundamental here. It just seems counter-intuitive.

Intriguing. I'll take a closer look and run more detailed experiments.

**Comment by ddemidov** [comment]
You should report the issue to the Boost.Compute developers here: https://github.com/boostorg/compute/issues.

**Comment by ddemidov** [comment]
There is another option you had from the very beginning. I am sorry for not suggesting it in the first place. [This](https://github.com/ddemidov/vexcl/blob/3343785848489020c23074e6a07414bd4ea6816d/vexcl/vector.hpp#L382-L394) vex::vector constructor takes an OpenCL/CUDA buffer and a size, and constructs a non-owning wrapper around that buffer.  It provides vexcl functionality for user-supplied buffers, but can easily be used in your case as well:
```cpp
for(int i = 0; i < n; i++ ) {
    vex::vector<int> keys_view(ctx.queue(0), keys(0), n - i);
    vex::vector<int> vals_view(ctx.queue(0), vals(0), n - i);

    vex::reduce_by_key(keys_view, vals_view, keys_out, vals_out, equal, maximum);
}
```

Note that you can not provide an offset into the buffer this way, but only its size.

**Comment by gcarvajalb** [comment]
Thanks for the suggestion.
I'll be away from my campus equipment until next week, but I'll try it as soon as I get back and report.

Cheers.

**Comment by gcarvajalb** [comment]
Ok. Following the recommendations I prepared a rough benchmark to estimate the performance gains of using the wrappers over my original naive implementation.

Using wrappers did work, but the performance gains where less than expected. Not saying that they are wrong or anything, it was mainly that I was being naively optimistic about getting a huge improvement for reducing the vector lengths on each iteration. After playing extensively, analysing the results, and giving it some thought, the results do make sense.

For example, I noticed that the difference in elapsed time of reduce_by_key with vectors of length between 1 and 1E6 elements (which is in the range of vectors I expect to operate with) was marginal, and there was no noticeable improvement in the total time of my algorithm. A figure summarizing this experiment is here: [exp1](https://www.dropbox.com/s/0v1s4xqs2ey1p8r/VexCL-reduce-by-key-time-from-1-to-1E6-elements.jpg?dl=0).  Note that the y-axis represents raw clock ticks (no transformation to seconds).

After scratching my head for a while, I moved to a newer GPU with more memory and tested with larger vectors of up to 10E6 elements. I noticed then that the performance gains start showing up with very large vectors, as shown here: [exp2]( https://www.dropbox.com/s/u8gx5iz2qdxsy2e/VexCL-reduce-by-key-time-from-50E3-to-10E6-elements.jpg?dl=0 ) , where reduce_by_key with 50K elements runs about 10x faster than with 10E6 elements.

Just to verify, the redefinitions of the vectors keys_view and vals_view in this loop:
```
for(int i = 0; i < n; i++ ) {
    vex::vector<int> keys_view(ctx.queue(0), keys(0), n - i);
    vex::vector<int> vals_view(ctx.queue(0), vals(0), n - i);

    vex::reduce_by_key(keys_view, vals_view, keys_out, vals_out, equal, maximum);
}
```
are just wrappesr? They do not copy or move any data, right?

In any case, the results make sense, and the exercise was incredibly helpful to understand some concepts I was omitting. I indeed detected other bottlenecks in my algorithm that may have a larger impact.

Again, thanks a lot for the prompt replies and useful tips. The library is turning out to be quite powerful (and hence useful).  For the record, the VexCL version of reduce_by_key still seems to be faster than the boost::compute version.

Just for completeness, the full code I used for testing is provided below. Hoepfully I'm not making a conceptual mistake there.

```
#include <numeric>
#include <iostream>
#include <vexcl/vector.hpp>
#include <vexcl/reduce_by_key.hpp>
#include <vexcl/function.hpp>

#define BODY_MAXIMUM          \
	if (x > y) return x; \
	else return y;

using namespace std;

const char getOffset_kernel_src[] = VEX_STRINGIZE_SOURCE(
    		kernel void getOffset(ulong n, __global int *trace, __global int *offset, int pivot, int resolution, ulong shift) 
    		{
      			const uint id = get_global_id(0);
    			if (id < n)                                             
        			offset[id] = (trace[id + shift] - pivot)/resolution;        
    		}
);

int main(int argc, char **argv)
{
	size_t N = 10000;
	int step = 1;
	int resolution = 10;

	if (argc >1)
	{
		N = atoi(argv[1]);
		step = atoi(argv[2]);
	}

	clock_t start, end;
	int i, j, counter, pivot;
	
	//generate synthetic input trace	
	vector<int> h_trace(N);

	h_trace[0] = 1234;
	srand (0); //set constant seed

	for(i=1; i<N; i++) // fill trace vector with increasing numbers
	{  	
		h_trace[i] = h_trace[i-1] + (rand() % 20 + 1);
	}

	// Init context for GPU
	vex::Context ctx(
		vex::Filter::Type(CL_DEVICE_TYPE_GPU) &&
		vex::Filter::Count(1)
		);

	if (ctx.queue().empty()) {
	    std::cerr << "GPU not found" << std::endl;
	    return 1;
	}

	//definition of functions and kernels
    	VEX_FUNCTION(bool, equal, (int, a)(int, b),
        	return (a == b);
        );

    	VEX_FUNCTION(int, maximum, (int, x)(int, y),
        BODY_MAXIMUM
        );

	vex::backend::kernel getOffset_kernel(ctx.queue(0), getOffset_kernel_src, "getOffset");

	//allocate vectors for computation
	vex::vector<int> d_trace(ctx, h_trace);
	vex::vector<int> d_offset(ctx, N);
	vex::vector<int> d_offset_complete(ctx, N);
    	vex::vector<int> d_okeys(ctx, N);
    	vex::vector<int> d_ovals(ctx, N);
	d_ovals = 0;

	std::vector<int> h_seq(N);	
	counter = 0;
	std::generate(h_seq.begin(), h_seq.end(), [&counter]{ return counter=counter+1;});
	vex::vector<int> d_seq(ctx, h_seq);	
	vex::copy(h_seq, d_seq);    	

	int num_keys;

	// call kernel here to remove compilation overhead from measured time in the loop (dummy call)
	num_keys = vex::reduce_by_key(
            			d_offset, d_seq,
            			d_okeys, d_ovals,
            			equal, maximum
        			);

	for(i= 0; i<N; i= i+step)
	{
		pivot = h_trace[i];
		//compute offset between pivot element and sucesive elements in the trace 
		getOffset_kernel(ctx.queue(0), static_cast<cl_ulong>(N-i), d_trace(0), d_offset_complete(0), pivot, resolution, static_cast<cl_ulong>(i));

		//allocate a new vector containing only the elements starting from pivot location
		vex::vector<int> d_offset_slice(ctx.queue(0), d_offset_complete(0), N-i);
    		vex::vector<int> d_seq_slice(ctx.queue(0), d_seq(0), N-i);
		vex::vector<int> d_okeys_slice(ctx.queue(0), d_okeys(0), N-i);
    		vex::vector<int> d_ovals_slice(ctx.queue(0), d_ovals(0), N-i);

		//measure time spent in reduce by key		
		start = clock();
 		num_keys = vex::reduce_by_key(
            			d_offset_slice, d_seq_slice, //d_offset, d_seq, //d_ikey, d_ivals,
            			d_okeys_slice, d_ovals_slice,
            			equal, maximum
        	);
		ctx.finish();
    		end = clock();
		std::cout << (d_offset_slice.end() - d_offset_slice.begin()) << " elements: " << (end - start) / (CLOCKS_PER_SEC / 1000000) << std::endl;
	}
}
```

**Comment by Reneg973** [comment]
Just a naive question, are you sure the queue has already processed all commands when calling start=clock() ?

**Comment by ddemidov** [comment]
> Just to verify, the redefinitions of the vectors keys_view and vals_view are just wrappers?

That is correct.

-----------------------------

```cpp
const char getOffset_kernel_src[] = VEX_STRINGIZE_SOURCE(
            kernel void getOffset(ulong n, __global int *trace, __global int *offset, int pivot, int resolution, ulong shift) 
            {
                const uint id = get_global_id(0);
                if (id < n)                                             
                    offset[id] = (trace[id + shift] - pivot)/resolution;        
            }
);
```

This kernel has a problem. For it to work correctly, you need to launch it on a grid with at least `n` threads. So you have to do something like [`getOffset_kernel.config((n+255)/256, 256)`](https://github.com/ddemidov/vexcl/blob/master/vexcl/backend/opencl/kernel.hpp#L223) after you compile the kernel. Or, you need to rewrite the kernel body to be
```cpp
                for(uint id = get_global_id(0); id < n; id += get_grid_size(0))
                    offset[id] = (trace[id + shift] - pivot)/resolution;        
```
That way it will work on grids of any size.

Even better, the kernel may be replaced with a simple expression:
```cpp
auto id = vex::tag<1>(vex::element_index(0, N-i));
vex::permutation(id)(d_offset) = (vex::permutation(id+shift)(d_trace) - pivot) / resolution;
```

-----------------------------

```cpp
    for(i=1; i<N; i++) // fill trace vector with increasing numbers
    {   
        h_trace[i] = h_trace[i-1] + (rand() % 20 + 1);
    }
    vex::vector<int> d_trace(ctx, h_trace);
```

Being pedantic here (and it should not matter for a benchmark like this), but this may be done on the GPU:
```cpp
vex::Random<int> irnd;
d_trace = 1234 + vex::element_index() + irnd(vex::element_index(), 0) % 20;
```

Same comment about initialization of `d_seq`.

-----------------------------

@Reneg973 has a valid point. You should `ctx.finish()` both before `start = clock()` and `end = clock()`. You could also use `vex::profiler<>` for that, see [examples/benchmark.cpp](https://github.com/ddemidov/vexcl/blob/master/examples/benchmark.cpp#L105-L109) for some use cases.

**Comment by ddemidov** [comment]
Actually, `benchmark.cpp` is a bad example of using `vex::profiler<>`. Profiler also has `tic_cl()` method which calls `ctx.finish()` before taking a measurement. `profiler::toc()` will do `ctx.finish()` if `tic_cl()` was called.

**Comment by gcarvajalb** [comment]
@Reneg973 , @ddemidov  : Thanks for the thorough revision and comments. I did make some mistakes in the code I posted, which is a skimmed version of the one I'm actually running in the computer (edited for simplicity).

I did overlook the call to `ctx.finish()` before starting the clock though, and in fact I had more kernels inside the loop. I corrected that and I see some changes in the measured times, but not that much.

Thanks a lot indeed. I feel now I've learned more OpenCL from these discussions than from the couple of books I read before jumping into it.

**Comment by gcarvajalb** [comment]
After a month of dealing with other deadlines, I jumped back to this as soon as I've got free and updated the code based on the feedback and suggestions.
Measuring the total execution time of the entire application (not only the dummy example I used), which makes heavy use of the loop under discussion, I've got a reduction in the running time between 30% and 40% for a very large vector after using the optimizations (1E6 + elements).
Neat!!

Thanks a lot for the responses and suggestions.
